lib/handler/proxy.c

```diff
@@ -60,38 +65,51 @@ static void on_context_init(h2o_handler_t *_self, h2o_context_t *ctx)
     /* use the loop of first context for handling socketpool timeouts */
     h2o_socketpool_register_loop(self->sockpool, ctx->loop);
 
+    struct rp_handler_context_t *handler_ctx = h2o_mem_alloc(sizeof(*handler_ctx));
+    memset(handler_ctx, 0, sizeof(*handler_ctx));
+    h2o_httpclient_connection_pool_init(&handler_ctx->connpool, self->sockpool);
+    h2o_context_set_handler_context(ctx, &self->super, handler_ctx);
+
     /* setup a specific client context only if we need to */
     if (ctx->globalconf->proxy.io_timeout == self->config.io_timeout &&
         ctx->globalconf->proxy.connect_timeout == self->config.connect_timeout &&
-        ctx->globalconf->proxy.first_byte_timeout == self->config.first_byte_timeout && !self->config.websocket.enabled)
+        ctx->globalconf->proxy.first_byte_timeout == self->config.first_byte_timeout &&
+        ctx->globalconf->proxy.keepalive_timeout == self->config.keepalive_timeout &&
+        ctx->globalconf->proxy.max_buffer_size == self->config.max_buffer_size &&
+        ctx->globalconf->proxy.http2.ratio == self->config.http2.ratio && !self->config.websocket.enabled)
         return;
 
-    h2o_http1client_ctx_t *client_ctx = h2o_mem_alloc(sizeof(*ctx));
-
+    h2o_httpclient_ctx_t *client_ctx = h2o_mem_alloc(sizeof(*ctx));
+    client_ctx->loop = ctx->loop;
+    client_ctx->getaddr_receiver = &ctx->receivers.hostinfo_getaddr;
     client_ctx->io_timeout = self->config.io_timeout;
     client_ctx->connect_timeout = self->config.connect_timeout;
     client_ctx->first_byte_timeout = self->config.first_byte_timeout;
-    client_ctx->loop = ctx->loop;
-    client_ctx->getaddr_receiver = &ctx->receivers.hostinfo_getaddr;
+    client_ctx->keepalive_timeout = self->config.keepalive_timeout;
     if (self->config.websocket.enabled) {
         client_ctx->websocket_timeout = &self->config.websocket.timeout;
     } else {
         client_ctx->websocket_timeout = NULL;
     }
 
-    h2o_context_set_handler_context(ctx, &self->super, client_ctx);
+    client_ctx->max_buffer_size = self->config.max_buffer_size;
+    client_ctx->http2.ratio = self->config.http2.ratio;
+    client_ctx->http2.counter = -1;
+
+    handler_ctx->client_ctx = client_ctx;
 }
 
 static void on_context_dispose(h2o_handler_t *_self, h2o_context_t *ctx)
 {
     struct rp_handler_t *self = (void *)_self;
-    h2o_http1client_ctx_t *client_ctx = h2o_context_get_handler_context(ctx, &self->super);
+    struct rp_handler_context_t *handler_ctx = h2o_context_get_handler_context(ctx, &self->super);
 
-    if (client_ctx == NULL)
-        return;
+    if (handler_ctx->client_ctx != NULL) {
+        h2o_socketpool_unregister_loop(self->sockpool, ctx->loop);
```

Comment:
This might be a bug in master, but would you mind looking into if we should call `h2o_socketpool_unregister_loop` regardless of the value of `client_ctx`?

Reply:
fixed in #1851.